lib/membrane_http_adaptive_stream/sink.ex

```diff
@@ -450,6 +435,59 @@ defmodule Membrane.HTTPAdaptiveStream.Sink do
     Storage.store_manifests(storage, manifest_files)
   end
 
+  defp total_partial_segments_duration(partials),
+    do: Enum.reduce(partials, 0, &(&1.metadata.duration + &2))
+
+  defp partial_segments_for_track(track_id, state) do
+    partial_segments = Map.get(state.track_to_partial_segments, track_id, [])
+
+    {partial_segments, total_partial_segments_duration(partial_segments)}
+  end
+
+  defp should_finalize_segment?(partial_buffer, partials, partials_duration, segment_duration) do
+    %{independent?: independent?, duration: duration} = partial_buffer.metadata
+
+    partials != [] and independent? and duration + partials_duration > segment_duration.min
+  end
+
+  defp finalize_segment_for_track(track_id, partials, partials_duration, manifest) do
+    payload =
+      partials
+      |> Enum.map(& &1.payload)
+      |> Enum.reverse()
+      |> IO.iodata_to_binary()
+
+    segment = %Membrane.Buffer{
+      payload: payload,
+      metadata: %{duration: partials_duration}
+    }
+
+    {changeset, manifest} = Manifest.finalize_current_segment(manifest, track_id)
+
+    {segment, changeset, manifest}
+  end
+
+  defp maybe_finalize_and_store_segment(track_id, state) do
+    %{manifest: manifest, storage: storage} = state
+
+    {partials, partials_duration} = partial_segments_for_track(track_id, state)
+
+    if partials != [] do
+      {segment, changeset, manifest} =
+        finalize_segment_for_track(track_id, partials, partials_duration, manifest)
+
+      case Storage.apply_segment_changeset(storage, track_id, changeset, segment) do
+        {:ok, storage} ->
+          {storage, manifest}
+
+        {{:error, reason}, _storage} ->
+          raise reason
+      end
+    else
+      {storage, manifest}
+    end
```

Comment:
```suggestion
    if partials == [] do
      {storage, manifest}
    else
      {segment, changeset, manifest} =
        finalize_segment_for_track(track_id, partials, partials_duration, manifest)

      case Storage.apply_segment_changeset(storage, track_id, changeset, segment) do
        {:ok, storage} -> {storage, manifest}
        {{:error, reason}, _storage} -> raise reason
      end
    end
```

Maybe it'll be a bit more readable this way?